### PR TITLE
Fluid benchmark support recordio reader

### PR DESCRIPTION
**Pull request by typhoonzero**
This can also fix the issue when running with `--gpus > 1`

### CODE REVIEW

File: benchmark/fluid/models/mnist.py [hunk 1]
```diff
@@ -65,9 +66,23 @@ def cnn_model(data):
 
 
 def get_model(args):
-    # Input data
-    images = fluid.layers.data(name='pixel', shape=[1, 28, 28], dtype=DTYPE)
-    label = fluid.layers.data(name='label', shape=[1], dtype='int64')
+    if args.use_reader_op:
+        filelist = [
+            os.path.join(args.data_path, f) for f in os.listdir(args.data_path)
```
Comment (Yancey1989) [inline benchmark/fluid/models/mnist.py]: We can use `glob` to specify the files.

File: benchmark/fluid/README.md [hunk 2]
```diff
@@ -42,6 +42,15 @@ Currently supported `--model` argument include:
     PADDLE_PSERVER_PORT=7164 PADDLE_TRAINER_IPS=192.168.0.2,192.168.0.3  PADDLE_CURRENT_IP=127.0.0.1 PADDLE_TRAINER_ID=0 python fluid_benchmark.py --model mnist --device GPU --update_method nccl2
     ```
 
+## Prepare the RecordIO file to Achieve Better Performance
+
+Run the following command will generate RecordIO files like "mnist.recordio" under the path
+and batch_size you choose:
+
+```bash
+python -c 'from recordio_converter import *; prepare_mnist("data", 32)'
```
Comment (Yancey1989) [inline benchmark/fluid/README.md]: It's better to set `batch_size=1`, we can set the `batch_size` in the trainer reader.

Comment (typhoonzero) [reply benchmark/fluid/README.md]: Done.

File: benchmark/fluid/fluid_benchmark.py [hunk 3]
```diff
@@ -205,26 +218,39 @@ def train(avg_loss, infer_prog, optimizer, train_reader, test_reader, batch_acc,
...
+            var for var in train_prog.global_block().vars.itervalues()
+            if var.is_data
+        ]
+        feeder = fluid.DataFeeder(feed_var_list, place)
 
     iters, num_samples, start_time = 0, 0, time.time()
     for pass_id in range(args.pass_num):
         train_losses = []
-        for batch_id, data in enumerate(train_reader()):
+        reader_generator = train_reader()
```
Comment (chengduoZH) [inline benchmark/fluid/fluid_benchmark.py]: `reader_generator = train_reader()` ==>
```
if not args.use_reader_op:
    reader_generator = train_reader()
```

File: benchmark/fluid/fluid_benchmark.py [hunk 4]
```diff
@@ -205,26 +218,39 @@ def train(avg_loss, infer_prog, optimizer, train_reader, test_reader, batch_acc,
...
+                loss = exe.run(train_prog, fetch_list=[avg_loss])
+            else:
+                loss = exe.run(train_prog,
+                               feed=feeder.feed(data),
+                               fetch_list=[avg_loss])
             iters += 1
-            num_samples += len(data)
+            batch_id += 1
+            # FIXME(wuyi): last batch size maybe different
+            num_samples += len(args.batch_size)
```
Comment (chengduoZH) [inline benchmark/fluid/fluid_benchmark.py]: For `use_reader_op`, if the current pass is not the last, the last batch of this pass is also equal to `args.batch_size`.

File: benchmark/fluid/fluid_benchmark.py [hunk 5]
```diff
@@ -282,12 +311,18 @@ def train_parallel(avg_loss, infer_prog, optimizer, train_reader, test_reader,
         num_trainers=num_trainers,
         trainer_id=trainer_id)
 
-    feeder = fluid.DataFeeder(feed_var_list, place)
     for pass_id in range(args.pass_num):
         num_samples = 0
         iters = 0
         start_time = time.time()
-        for batch_id, data in enumerate(train_reader()):
+        reader_generator = train_reader()
```
Comment (chengduoZH) [inline benchmark/fluid/fluid_benchmark.py]: `reader_generator = train_reader()` ==>
```
if not args.use_reader_op:
    reader_generator = train_reader()
```

File: benchmark/fluid/models/mnist.py [hunk 6]
```diff
@@ -65,9 +66,23 @@ def cnn_model(data):
...
+        ]
+        data_file = fluid.layers.open_files(
+            filenames=filelist,
+            shapes=[[-1, 1, 28, 28], (-1, 1)],
+            lod_levels=[0, 0],
+            dtypes=["float32", "int64"],
+            thread_num=args.gpus)
+        data_file = fluid.layers.double_buffer(
+            fluid.layers.batch(
+                data_file, batch_size=args.batch_size))
```
Comment (chengduoZH) [inline benchmark/fluid/models/mnist.py]: For `use_reader_op`, the `batch_size` of `fluid.layers.batch` is set with a single card, this is to say if the batch size is 256 when training Vgg, and the machine has 4 cards, the `batch_size` for `fluid.layers.batch` should be 64.

File: benchmark/fluid/models/vgg.py [hunk 7]
```diff
@@ -65,9 +66,23 @@ def get_model(args):
...
+        ]
+        data_file = fluid.layers.open_files(
+            filenames=filelist,
+            shapes=[[-1] + data_shape, (-1, 1)],
+            lod_levels=[0, 0],
+            dtypes=["float32", "int64"],
+            thread_num=args.gpus)
+        data_file = fluid.layers.double_buffer(
+            fluid.layers.batch(
+                data_file, batch_size=args.batch_size))
```
Comment (chengduoZH) [inline benchmark/fluid/models/vgg.py]: Same as above.

File: benchmark/fluid/fluid_benchmark.py [hunk 8]
```diff
@@ -296,9 +331,10 @@ def train_parallel(avg_loss, infer_prog, optimizer, train_reader, test_reader,
             if iters == args.skip_batch_num:
                 start_time = time.time()
                 num_samples = 0
-            if iters == args.iterations:
+            # NOTE: if use reader ops, the input data is not splited to multiple cards
+            if args.use_reader_op and iters >= args.iterations / args.gpus:
```
Comment (chengduoZH) [inline benchmark/fluid/fluid_benchmark.py]: I don't think `iters >= args.iterations / args.gpus` is appropriate. 
Because the model's accuracy is highly related to the new parameters that have learned, but the new parameters may be related to the times of updating parameter. So maybe we should not do that.

Comment (typhoonzero) [reply benchmark/fluid/fluid_benchmark.py]: Well, `args.iterations` is intended to let the benchmark finish fast, no concerns for model accuracy. To run a full model training, we can set `args.iterations` to -1 so that it can run until all train data have been fed.

File: benchmark/fluid/fluid_benchmark.py [hunk 9]
```diff
@@ -266,7 +266,10 @@ def train(avg_loss, infer_prog, optimizer, train_reader, test_reader, batch_acc,
             # FIXME(wuyi): For use_reader_op, if the current
             # pass is not the last, the last batch of this pass
             # is also equal to args.batch_size.
-            num_samples += len(args.batch_size)
+            if args.use_reader_op:
+                num_samples += args.batch_size
```
Comment (chengduoZH) [inline benchmark/fluid/fluid_benchmark.py]: `args.batch_size` is the batch size on each GPU now. So it should be ` num_samples += args.batch_size * args.gpus`

Comment (typhoonzero) [reply benchmark/fluid/fluid_benchmark.py]: Thanks very much! Done. Current know issue, if set `--use_reader_op` we must also set `--no_test` will fix this in next PR.
